url3472127e1f_SergeyKrivohatskiy_tiny_database_tgz-pJ8-ru_spbau_tinydb_bufferManager_IndexTestJ8/src/ru/spbau/tinydb/grammar/SQLGrammarParser.java

```diff
@@ -0,0 +1,3083 @@
+			if ( listener instanceof SQLGrammarListener ) ((SQLGrammarListener)listener).exitValue(this);
+		}
+	}
+
+	public final ValueContext value() throws RecognitionException {
+		ValueContext _localctx = new ValueContext(_ctx, getState());
+		enterRule(_localctx, 12, RULE_value);
+		try {
+			setState(148);
+			switch ( getInterpreter().adaptivePredict(_input,4,_ctx) ) {
+			case 1:
+				enterOuterAlt(_localctx, 1);
+				{
+				setState(139); ((ValueContext)_localctx).integerLiteral = integerLiteral();
+
+				        ((ValueContext)_localctx).result =  ((ValueContext)_localctx).integerLiteral.result;
+				    
+				}
+				break;
+			case 2:
+				enterOuterAlt(_localctx, 2);
+				{
+				setState(142); ((ValueContext)_localctx).doubleLiteral = doubleLiteral();
+
+				        ((ValueContext)_localctx).result =  ((ValueContext)_localctx).doubleLiteral.result;
+				    
+				}
+				break;
+			case 3:
+				enterOuterAlt(_localctx, 3);
+				{
+				setState(145); ((ValueContext)_localctx).varCharLiteral = varCharLiteral();
+
+				        ((ValueContext)_localctx).result =  ((ValueContext)_localctx).varCharLiteral.result;
+				    
+				}
+				break;
+			}
+		}
+		catch (RecognitionException re) {
+			_localctx.exception = re;
+			_errHandler.reportError(this, re);
+			_errHandler.recover(this, re);
+		}
+		finally {
+			exitRule();
+		}
+		return _localctx;
+	}
+
+	public static class IntegerLiteralContext extends ParserRuleContext {
+		public Integer result;
+		public List<TerminalNode> DECIMAL_DIGIT() { return getTokens(SQLGrammarParser.DECIMAL_DIGIT); }
+		public TerminalNode SIGN() { return getToken(SQLGrammarParser.SIGN, 0); }
+		public TerminalNode DECIMAL_DIGIT(int i) {
+			return getToken(SQLGrammarParser.DECIMAL_DIGIT, i);
+		}
+		public IntegerLiteralContext(ParserRuleContext parent, int invokingState) {
+			super(parent, invokingState);
+		}
+		@Override public int getRuleIndex() { return RULE_integerLiteral; }
+		@Override
+		public void enterRule(ParseTreeListener listener) {
+			if ( listener instanceof SQLGrammarListener ) ((SQLGrammarListener)listener).enterIntegerLiteral(this);
+		}
+		@Override
+		public void exitRule(ParseTreeListener listener) {
+			if ( listener instanceof SQLGrammarListener ) ((SQLGrammarListener)listener).exitIntegerLiteral(this);
+		}
+	}
+
+	public final IntegerLiteralContext integerLiteral() throws RecognitionException {
+		IntegerLiteralContext _localctx = new IntegerLiteralContext(_ctx, getState());
+		enterRule(_localctx, 14, RULE_integerLiteral);
+		int _la;
+		try {
+			enterOuterAlt(_localctx, 1);
+			{
+			setState(151);
+			_la = _input.LA(1);
+			if (_la==SIGN) {
+				{
+				setState(150); match(SIGN);
+				}
+			}
+
+			setState(154); 
+			_errHandler.sync(this);
+			_la = _input.LA(1);
+			do {
+				{
+				{
+				setState(153); match(DECIMAL_DIGIT);
+				}
+				}
+				setState(156); 
+				_errHandler.sync(this);
+				_la = _input.LA(1);
+			} while ( _la==DECIMAL_DIGIT );
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 649:658, 719:728, 733:742, 766:775.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.